### PR TITLE
✨ added lookalike functionality

### PR DESCRIPTION
**Pull request by SamBajanik**

### CODE REVIEW

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Functions
+
+# COMMAND ----------
+
+def define_lookalikes(value, df_predictions):
+    if (isinstance(value, float)):
```
Comment (guderkar) [inline odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:36]: what is the purpose if this if else? if float then fiter, if int then limit? Shouldn't we split it in two different functions?

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        report_string = f"Computation completed. The lowest probability of the lookalike subject admitted: {lowest_prob}"
+    
+    print(report_string)
+
+    return df_lookalikes
+
+# COMMAND ----------
+
+def get_date_parts(date_string):
+    return [int(date_part) for date_part in date_string.split("-")]
```
Comment (guderkar) [inline odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:63]: can't we just parse with datetime module? and return datitime object, then access like 
```
parsed_date = parse_date(date_str)
parsed_date.year/month/day
```

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+# COMMAND ----------
+
+def predict(features: DataFrame, model: PipelineModel, feature_name: str):
+    predictions_df = model.transform(features)
```
Comment (guderkar) [inline odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:68]: why we are aliasing predicted column by feature_name?

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+    if "probability" in predictions_df.columns:
```
Comment (guderkar) [inline odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:70]: not sure why sometimes it's probability and sometimes prediction but this code can be refactored like this
```
prediction_col = f.round(f.element_at(vector_to_array(f.col("probability")), 2).cast("float"), 3) if "probability" in predictions_df.columns else f.round("prediction", 3)

return predictions_df.select(
    dbutils.widgets.get("entity_id_column_name"),
    "timestamp",
    prediction_col.alias(feature_name),
)
```

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+df_to_enrich = (
+    spark.table("odap_digi_use_case_segments.segments")
+    .select(dbutils.widgets.get("entity_id_column_name"))
+    .filter(f.col("segment") == dbutils.widgets.get("segment_name"))
+)
+
+df_inference_dataset = df_data.join(
+    df_to_enrich,
+    on=dbutils.widgets.get("entity_id_column_name"),
+    how="anti",
```
Comment (guderkar) [inline odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:119]: is it left anti join? if yes write it as leftanti

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+).withColumn("label", f.lit(0))
+
+df_modelling_dataset = df_data.join(
+    df_to_enrich,
+    on=dbutils.widgets.get("entity_id_column_name"),
+    how="semi",
```
Comment (guderkar) [inline odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:125]: leftsemi?

File: odap-framework-demo/use_cases/upsell_uc/segments/lals_cookies_highly_likely_to_submit.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        df_output = predictions_df.select(
+            entity_id_column_name,
+            "timestamp",
+            f.round("prediction", 3).alias(feature_name),
+        )
+    return df_output
+
+# COMMAND ----------
+
+def return_slider(criterion):
```
Comment (guderkar) [inline odap-framework-demo/use_cases/upsell_uc/segments/lals_cookies_highly_likely_to_submit.py:88]: do we need the slider here?

**Review by guderkar — CHANGES_REQUESTED** [review]

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 1 above)
Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:36]: it is a function that should respond to the choices of the user he makes interactively, he can choose two criterions for defining lookalikes and based on the choice, we must either use filter or limit logic, the output of the function is always the lookalike segment, therefore I decided to put it into one function

File: odap-framework-demo/use_cases/upsell_uc/segments/lals_cookies_highly_likely_to_submit.py (hunk 7 above)
Comment (SamBajanik) [reply odap-framework-demo/use_cases/upsell_uc/segments/lals_cookies_highly_likely_to_submit.py:88]: you are right, we do not need it, i will remove it

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 5 above)
Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:119]: yes, left

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 6 above)
Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:125]: yes, left

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 4 above)
Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py]: thanks, i will replace it with this code

Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py]: sometimes, the model output does not have to contain probability colum

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 1 above)
Comment (guderkar) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:36]: if those functions are moved in plain .py files and imported then would it work?

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 2 above)
Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py]: ok, I have modified it this way

**Review by guderkar — CHANGES_REQUESTED** [review]

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 3 above)
Comment (guderkar) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py]: but feature is kinda weird name for this variable. Shouldn't it be something like `prediction_output_col_name` with default value of `"prediction"`?

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 1 above)
Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py:36]: yes, i think so

File: odap-framework-demo/lookalike_modelling/explore_the_lookalike.py (hunk 3 above)
Comment (SamBajanik) [reply odap-framework-demo/lookalike_modelling/explore_the_lookalike.py]: it is taken from ai features, thats it is called so, but i will rename it for this purpose, you are right

**Review by guderkar — APPROVED** [review]

**Pull request merged**